_posts/2020-04-04-software-development-best-practices.md

```diff
@@ -0,0 +1,70 @@
+tags:
+- PostgreSQL
+- DDoS
+thumbnail_path: blog/thumbs/software-development-best-practices.png
+add_to_popular_list: true
+---
+
+Well-written project is more stable, easier to maintain, more enjoyable to work. In the following post, we will discuss some best practices that will make your project take your project to a new level.
+
+{% include figure.html path=page.thumbnail_path %}
+
+## Best Practices
+
+#### Deliver Working Software
+
+**[Pair programming](#pair-programming)** is development technique in which two programmers work on a specific problem together.
+
+**[Test-driven development](#test-driven-development)** improves code design and quality.
+
+#### Work Together
+
+**[Scrum meetings](#scrum-meetings)** are short meetings used to plan, review, and increase accountability and clarity across a team.
+
+**[Retrospectives](#retrospectives)** a period for reflection at the end of each sprint.
+
+
+## Pair programming
+
+{% include figure.html path=page.thumbnail_path path="blog/software-development-best-practices/pair-prog.png" %}
+
+Pair programming is an agile software development technique in which two programmers work together at one workstation. One, the driver, writes code while the other, the observer or navigator, reviews each line of code as it is typed in. The two programmers switch roles frequently.
+
+A system with two programmers possesses greater potential for the generation of more diverse solutions to problems for three reasons:
+
+1. the programmers bring different prior experiences to the task;
+2. they may assess information relevant to the task in different ways;
+3. they stand in different relationships to the problem by virtue of their functional roles.
+
+## Test-driven development
+
+Test driven development will improve the design and quality of the code written. It follows the cycle:
+
+{% include figure.html path=page.thumbnail_path path="blog/software-development-best-practices/tdd.png" %}
+
+## Scrum meetings
+
+{% include figure.html path=page.thumbnail_path path="blog/software-development-best-practices/scrum.png" %}
+
+Scrum is when a team quickly groups up in an effort to "self-organize". It is recommended to keep the scrum as short as possible. Agree on a duration of about 5-15 minutes. Scrums are typically conducted at the very beginning of every day additionally as needed (i.e. afternoon and the end of the day). A standard format is to go around in a circle and for each person to answer the following questions:
+
+* What user stories did you complete yesterday?
+* What user stories are you working on today?
+* Do you have any blocks?
+
+## Retrospectives
+
+{% include figure.html path=page.thumbnail_path path="blog/software-development-best-practices/retro.jpg" %}
+
+At the end of each sprint a team will conduct a retrospective in an effort to "become more effective". There are numerous formats for which to conduct a retrospective; the six hats is a popular one.
```

Comment:
1. why `become more effective` is in ""?
2. What are the six hats? It's not obvious

Reply:
At the end of each sprint,